docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+    # resident in HBM.
+    jax.ShapeDtypeStruct((2, 8, 128), jnp.float32),  # hbm_scratch
+)
+
+grid_spec = pltpu.PrefetchScalarGridSpec(
+    num_scalar_prefetch=0,
+    in_specs=[
+        # Our input lives in VMEM
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.VMEM),
+    ],
+    out_specs=[
+        # Our output lives in VMEM
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.VMEM),
+        # Our double-buffer lives in HBM
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+    ],
+    grid=(num_devices,),
+    scratch_shapes=(
+        [pltpu.SemaphoreType.DMA] * 3
+        + [pltpu.SemaphoreType.REGULAR]  # capacity_sem
+        + [pltpu.VMEM((8, 128), jnp.float32)]  # receive_scratch
+    ),
+)
+
+kernel = pl.pallas_call(
+    all_reduce_kernel,
+    out_shape=out_shape,
+    grid_spec=grid_spec,
+    compiler_params=dict(mosaic=dict(collective_id=0)),
+)
+
+pallas_result = jax.jit(
+    shard_map.shard_map(
+        kernel,
+        mesh=mesh,
+        in_specs=partition,
+        out_specs=partition,
+        check_rep=False,
+    )
+)(input_arr)
+pallas_result = jax.block_until_ready(pallas_result)[0]
+
+
+def lax_sum(x):
+  return lax.psum(x, 'x')
+
+
+xla_result = jax.jit(
+    shard_map.shard_map(
+        lax_sum, mesh=mesh, in_specs=P(None, 'x'), out_specs=P(None, 'x')
+    )
+)(input_arr)
+
+print('Input = ', input_arr[0, ::128])
+print('Pallas result = ', pallas_result[0, ::128])
+print('lax.psum result = ', xla_result[0, ::128])
+difference = jnp.mean(jnp.abs(pallas_result - xla_result))
+print('Difference |Pallas - lax.psum| = ', difference)
+```
+
++++ {"id": "d8bsZAzQreC_"}
+
+### Run-ahead and Race Conditions
+
+As a general rule of thumb, to maximize performance we want to allow a device to run-ahead of other devices without synchronization as much as possible without sacrificing correctness of the program. While we could enforce a barrier across all devices at the beginning of each iteration, this bottlenecks the performance of the program to the slowest device on each loop. By relaxing synchronization and allowing a moderate amount of run-ahead, we can better accommodate variance in latency between iterations and devices because a device that is slow on one iteration could catch up on the next iteration.
+
+In the all-reduce kernel we wrote previously, we allow devices to run ahead but by less than one iteration compared to its neighbors (however, non-neighboring devices could be more than 1 iteration apart). To see why the semaphore synchronization is necessary, consider the case when one device (say device 2) hangs and falls behind the other devices. Waiting on a DMA is one-directional - it only blocks the receiver and not the sender. Therefore, each device can run up to one iteration ahead before it becomes blocked waiting for the next DMA to arrive. If we have N devices, this means that the final device can be up to N iterations ahead of the first device.
```

Comment:
```suggestion
In the all-reduce kernel we wrote previously, we allow devices to run ahead but by less than one iteration compared to its neighbors (however, non-neighboring devices could be more than 1 iteration apart). To see why the semaphore synchronization is necessary, consider the case when one device (say device 2) hangs and falls behind the other devices. Waiting on a DMA is one-directional — it only blocks the receiver and not the sender. Therefore, each device can run up to one iteration ahead before it becomes blocked waiting for the next DMA to arrive. If we have N devices, this means that the final device can be up to N iterations ahead of the first device.
```

Don't send a hyphen to do an m-dash's job :P

Reply:
Done.